File: iOS/FitGAI/PoseLogic/Classification/RepetitionCounter.swift
//
//  RepetitionCounter.swift
//  FitGAI
//
//  Created by 김기훈 on 2021/12/13.
//

import Foundation

class RepetitionCounter: NSObject {
    static let DEFAULT_ENTER_THRESHOLD:Float = 6.0
    static let DEFAULT_EXIT_THRESHOLD:Float = 4.0
    let enterThreshold: Float = RepetitionCounter.DEFAULT_ENTER_THRESHOLD
    let exitThreshold: Float = RepetitionCounter.DEFAULT_EXIT_THRESHOLD

    var className:String
    var numRepeats = 0
    var poseEntered = false

    init(className:String) {
        self.className = className
        super.init()
    }
    
    func addClassificationResult(classificationResult:ClassificationResult) -> Int {
        let poseConfident = classificationResult.getClassConfidence(className: className)
        if !poseEntered {
            poseEntered = poseConfident > RepetitionCounter.DEFAULT_ENTER_THRESHOLD
            return numRepeats
        }
        
        if poseConfident < exitThreshold {
            numRepeats += 1
            poseEntered = false
        }
        return numRepeats
    }
    
}

File: Users/jeong-yeongjae/sub_tree_test/presentation/src/main/java/com/hnh/presentation/logic/classification/RepetitionCounter.kt
package com.hnh.presentation.logic.classification

class RepetitionCounter @JvmOverloads constructor(
    val className: String,
    private val enterThreshold: Float = DEFAULT_ENTER_THRESHOLD,
    private val exitThreshold: Float = DEFAULT_EXIT_THRESHOLD
) {
    var numRepeats = 0
        private set
    private var poseEntered = false

    fun addClassificationResult(classificationResult: ClassificationResult): Int {
        val poseConfidence = classificationResult.getClassConfidence(className)
        if (!poseEntered) {
            poseEntered = poseConfidence > enterThreshold
            return numRepeats
        }
        if (poseConfidence < exitThreshold) {
            numRepeats++
            poseEntered = false
        }
        return numRepeats
    }
    @JvmName("getClassName1")
    fun getClassName(): String {
        return className
    }

    @JvmName("getNumRepeats1")
    fun getNumRepeats(): Int {
        return numRepeats
    }
    companion object {
        // These thresholds can be tuned in conjunction with the Top K values in {@link PoseClassifier}.
        // The default Top K value is 10 so the range here is [0-10].
        private const val DEFAULT_ENTER_THRESHOLD = 6f
        private const val DEFAULT_EXIT_THRESHOLD = 4f
    }
}